File: problem167/two_sum.go
package problem167

func twoSum(numbers []int, target int) []int {
	for i := len(numbers) - 1; i >= 0; i-- {
		if numbers[i] <= target-numbers[0] {
			j := 0
			for j < i {
				if numbers[j]+numbers[i] == target {
					return []int{j + 1, i + 1}
				} else if numbers[j]+numbers[i] < target {
					j++
				} else {
					i--
				}
			}
		}
	}
	return nil
}

File: two_sum.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

class Solution:
    def twoSum(self, nums, target):
        numMap = {}
        for i in range(len(nums)):
            if (target - nums[i]) in numMap:
                return [i, numMap[target - nums[i]]]
            else:
                numMap[nums[i]] = i

if __name__ == '__main__':
    assert Solution().twoSum([3, 2, 4], 6) == [1, 2]
